go-agent 23.1.0 does not start when system tmp dir has a noexec mount

##### Issue Type

Bug report

##### Summary

v23.1.0 Go-Server installed in an instance and working fine
v23.1.0 Go-Agent installed in different instance and service can't start because of Java error

##### Environment


###### Basic environment details

```
[Thu May 11 08:49:31 UTC 2023] Starting process:
[Thu May 11 08:49:31 UTC 2023]   Working directory    : /var/lib/go-agent
[Thu May 11 08:49:31 UTC 2023]   Application arguments: [-serverUrl, http://<go-server_ip_addr>:8153/go, -sslVerificationMode, FULL]
[Thu May 11 08:49:31 UTC 2023]            GoCD Version: 23.1.0-16079
[Thu May 11 08:49:31 UTC 2023]            Java Version: 17.0.2
[Thu May 11 08:49:31 UTC 2023]        Operating System: Linux(4.18.0-348.20.1.el7.aarch64)
```

###### Additional Environment Details


##### Steps to Reproduce

Assuming that we have installed Go-server and everything is fine. Now we need to prepare the go-agent server:

1. Installing Java17 as it is required by go-agent:
```
wget https://download.java.net/java/GA/jdk17.0.2/dfd4a8d0985749f896bed50d7138ee7f/8/GPL/openjdk-17.0.2_linux-aarch64_bin.tar.gz
tar xvf openjdk-17.0.2_linux-aarch64_bin.tar.gz && mv jdk-17.0.2/ /opt/jdk-17/
```

2. Installing the go-agent latest version (23.1.0):
```
curl https://download.gocd.org/gocd.repo -o /etc/yum.repos.d/gocd.repo
yum install go-agent -y
```

3. Installing Java11 in the system as default version:
```
yum install java-11-openjdk java-11-openjdk-devel java-11-openjdk-headless
```

4. Replacing amd64 bits with arm64 ones we have had in the system:
```
rm -f /usr/share/go-agent/jre/bin/java /usr/share/go-agent/jre/bin/jfr /usr/share/go-agent/jre/bin/jrunscript /usr/share/go-agent/jre/bin/keytool /usr/share/go-agent/jre/bin/rmiregistry
ln -s /bin/jfr /usr/share/go-agent/jre/bin/jfr
ln -s /bin/jrunscript /usr/share/go-agent/jre/bin/jrunscript
ln -s /bin/keytool /usr/share/go-agent/jre/bin/keytool
ln -s /bin/rmiregistry /usr/share/go-agent/jre/bin/rmiregistry
ln -s /opt/jdk-17/bin/java /usr/share/go-agent/jre/bin/java
```

5. Filling out the wrapper-properties:
```
echo > /usr/share/go-agent/wrapper-config/wrapper-properties.conf <<EOF
wrapper.app.parameter.100=-serverUrl
wrapper.app.parameter.101=http://<go-server-IP-addr>:8153/go
EOF
```

6. Running the go-agent
```
systemctl start go-agent
```

##### Expected Results
Go-agent starts and appears in the agent list of the server

##### Actual Results
Go-Agent couldn't start properly and never appears in the go-server

##### Possible Fix

##### Log snippets
```
==> /var/log/go-agent/go-agent-stderr.log <==
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at com.thoughtworks.gocd.Boot.run(Boot.java:90)
	at com.thoughtworks.gocd.Boot.main(Boot.java:56)
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'agentController' defined in class path resource [applicationContext.xml]: Invocation of init method failed; nested exception is java.lang.NoClassDefFoundError: Could not initialize class com.sun.jna.Native
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1619)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:553)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:481)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:312)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:308)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:756)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:867)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:542)
	at org.springframework.context.support.ClassPathXmlApplicationContext.<init>(ClassPathXmlApplicationContext.java:139)
	at org.springframework.context.support.ClassPathXmlApplicationContext.<init>(ClassPathXmlApplicationContext.java:83)
	at com.thoughtworks.go.agent.AgentMain.main(AgentMain.java:43)
	... 6 more
Caused by: java.lang.NoClassDefFoundError: Could not initialize class com.sun.jna.Native
	at com.sun.jna.NativeLong.<clinit>(NativeLong.java:35)
	at oshi.util.FileUtil.readNativeLongFromBuffer(FileUtil.java:208)
	at oshi.driver.linux.proc.Auxv.queryAuxv(Auxv.java:39)
	at oshi.software.os.linux.LinuxOperatingSystem.<clinit>(LinuxOperatingSystem.java:87)
	at oshi.SystemInfo.createOperatingSystem(SystemInfo.java:85)
	at oshi.util.Memoizer$1.get(Memoizer.java:61)
	at oshi.SystemInfo.getOperatingSystem(SystemInfo.java:76)
	at com.thoughtworks.go.agent.service.SystemInfo.determineOperatingSystemCompleteName(SystemInfo.java:36)
	at com.thoughtworks.go.agent.service.SystemInfo.<clinit>(SystemInfo.java:28)
	at com.thoughtworks.go.server.service.AgentRuntimeInfo.refreshOperatingSystem(AgentRuntimeInfo.java:200)
	at com.thoughtworks.go.server.service.AgentRuntimeInfo.fromAgent(AgentRuntimeInfo.java:73)
	at com.thoughtworks.go.agent.AgentController.initRuntimeInfo(AgentController.java:167)
	at com.thoughtworks.go.agent.AgentController.init(AgentController.java:135)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeCustomInitMethod(AbstractAutowireCapableBeanFactory.java:1748)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1685)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1615)
	... 18 more
```


##### Code snippets/Screenshots

##### Any other info

I've tried this on new instance:
```
$ yum install https://download.oracle.com/java/17/latest/jdk-17_linux-aarch64_bin.rpm -y
$ java --version
java 17.0.7 2023-04-18 LTS
Java(TM) SE Runtime Environment (build 17.0.7+8-LTS-224)
Java HotSpot(TM) 64-Bit Server VM (build 17.0.7+8-LTS-224, mixed mode, sharing)
$ curl https://download.gocd.org/gocd.repo -o /etc/yum.repos.d/gocd.repo && yum install go-agent -y
$ cat > /usr/share/go-agent/wrapper-config/wrapper-properties.conf <<EOF
wrapper.app.parameter.100=-serverUrl
wrapper.app.parameter.101=http://<go-server-ip-addr>:8153/go
wrapper.java.command=/usr/lib/jvm/jdk-17-oracle-aarch64/bin/java
EOF
$ systemctl start go-agent
```

Now I am having just java 17 on the system and the error is same:
```
==> /var/log/go-agent/go-agent-stderr.log <==
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at com.thoughtworks.gocd.Boot.run(Boot.java:90)
	at com.thoughtworks.gocd.Boot.main(Boot.java:56)
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'agentController' defined in class path resource [applicationContext.xml]: Invocation of init method failed; nested exception is java.lang.NoClassDefFoundError: Could not initialize class com.sun.jna.Native
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1619)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:553)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:481)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:312)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:308)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:756)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:867)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:542)
	at org.springframework.context.support.ClassPathXmlApplicationContext.<init>(ClassPathXmlApplicationContext.java:139)
	at org.springframework.context.support.ClassPathXmlApplicationContext.<init>(ClassPathXmlApplicationContext.java:83)
	at com.thoughtworks.go.agent.AgentMain.main(AgentMain.java:43)
	... 6 more
Caused by: java.lang.NoClassDefFoundError: Could not initialize class com.sun.jna.Native
	at com.sun.jna.NativeLong.<clinit>(NativeLong.java:35)
	at oshi.util.FileUtil.readNativeLongFromBuffer(FileUtil.java:208)
	at oshi.driver.linux.proc.Auxv.queryAuxv(Auxv.java:39)
	at oshi.software.os.linux.LinuxOperatingSystem.<clinit>(LinuxOperatingSystem.java:87)
	at oshi.SystemInfo.createOperatingSystem(SystemInfo.java:85)
	at oshi.util.Memoizer$1.get(Memoizer.java:61)
	at oshi.SystemInfo.getOperatingSystem(SystemInfo.java:76)
	at com.thoughtworks.go.agent.service.SystemInfo.determineOperatingSystemCompleteName(SystemInfo.java:36)
	at com.thoughtworks.go.agent.service.SystemInfo.<clinit>(SystemInfo.java:28)
	at com.thoughtworks.go.server.service.AgentRuntimeInfo.refreshOperatingSystem(AgentRuntimeInfo.java:200)
	at com.thoughtworks.go.server.service.AgentRuntimeInfo.fromAgent(AgentRuntimeInfo.java:73)
	at com.thoughtworks.go.agent.AgentController.initRuntimeInfo(AgentController.java:167)
	at com.thoughtworks.go.agent.AgentController.init(AgentController.java:135)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeCustomInitMethod(AbstractAutowireCapableBeanFactory.java:1748)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1685)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1615)
	... 18 more
Caused by: java.lang.ExceptionInInitializerError: Exception java.lang.UnsatisfiedLinkError: /tmp/jna434117078481342640.tmp: /tmp/jna434117078481342640.tmp: failed to map segment from shared object: Operation not permitted [in thread "main"]
	at java.base/jdk.internal.loader.NativeLibraries.load(Native Method)
	at java.base/jdk.internal.loader.NativeLibraries$NativeLibraryImpl.open(NativeLibraries.java:388)
	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:232)
	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:174)
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2389)
	at java.base/java.lang.Runtime.load0(Runtime.java:755)
	at java.base/java.lang.System.load(System.java:1953)
	at com.sun.jna.Native.loadNativeDispatchLibraryFromClasspath(Native.java:1045)
	at com.sun.jna.Native.loadNativeDispatchLibrary(Native.java:1015)
	at com.sun.jna.Native.<clinit>(Native.java:221)
	at com.sun.jna.platform.linux.Udev.<clinit>(Udev.java:37)
	at oshi.software.os.linux.LinuxOperatingSystem.<clinit>(LinuxOperatingSystem.java:74)
	... 34 more
```

And I will add other log files, as maybe there's something useful that I am missing to mention:
```
==> /var/log/go-agent/go-agent-stderr.log <==
[Thu May 11 13:30:41 UTC 2023] Starting process:
[Thu May 11 13:30:41 UTC 2023]   Working directory    : /var/lib/go-agent
[Thu May 11 13:30:41 UTC 2023]   Application arguments: [-serverUrl, http://<go-server_ip_addr>:8153/go, -sslVerificationMode, FULL]
[Thu May 11 13:30:41 UTC 2023]            GoCD Version: 23.1.0-16079
[Thu May 11 13:30:41 UTC 2023]            Java Version: 17.0.7
[Thu May 11 13:30:41 UTC 2023]        Operating System: Linux(4.18.0-348.20.1.el7.aarch64)

Could not find file `config/agent-logback.xml'. Attempting to load from classpath.
Using classpath resource `jar:onejar:lib/agent-23.1.0-16079-classes.jar!/config/agent-logback.xml'.
```

```
==> /var/log/go-agent/go-agent.log <==
2023-05-11 13:30:42,392 INFO  [main] ClassPathXmlApplicationContext:583 - Refreshing org.springframework.context.support.ClassPathXmlApplicationContext@72057ecf: startup date [Thu May 11 13:30:42 UTC 2023]; root of context hierarchy

2023-05-11 13:30:42,443 INFO  [main] XmlBeanDefinitionReader:317 - Loading XML bean definitions from class path resource [applicationContext.xml]
2023-05-11 13:30:42,707 INFO  [main] XmlBeanDefinitionReader:317 - Loading XML bean definitions from class path resource [applicationContext-plugin-infra.xml]
2023-05-11 13:30:43,010 INFO  [main] DefaultListableBeanFactory:821 - Overriding bean definition for bean 'pluginLoader' with a different definition: replacing [Generic bean: class [com.thoughtworks.go.plugin.infra.PluginLoader]; scope=singleton; abstract=false; lazyInit=false; autowireMode=0; dependencyCheck=0; autowireCandidate=true; primary=false; factoryBeanName=null; factoryMethodName=null; initMethodName=null; destroyMethodName=null; defined in URL [jar:onejar:lib/go-plugin-infra-23.1.0-16079.jar!/com/thoughtworks/go/plugin/infra/PluginLoader.class]] with [Generic bean: class [com.thoughtworks.go.plugin.infra.PluginLoader]; scope=; abstract=false; lazyInit=false; autowireMode=0; dependencyCheck=0; autowireCandidate=true; primary=false; factoryBeanName=null; factoryMethodName=null; initMethodName=null; destroyMethodName=null; defined in class path resource [applicationContext-plugin-infra.xml]]
2023-05-11 13:30:43,275 INFO  [main] AutowiredAnnotationBeanPostProcessor:156 - JSR-330 'javax.inject.Inject' annotation found and supported for autowiring
2023-05-11 13:30:43,503 INFO  [main] AgentStatusHttpd:88 - Agent status HTTP API server running on http://localhost:8152.
2023-05-11 13:30:45,276 WARN  [main] ClassPathXmlApplicationContext:550 - Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'agentController' defined in class path resource [applicationContext.xml]: Invocation of init method failed; nested exception is java.lang.NoClassDefFoundError: Could not initialize class com.sun.jna.Native
```

```
==> /var/log/go-agent/go-agent-bootstrapper.log <==
2023-05-11 13:32:21,611 INFO  [WrapperJarAppMain] AgentBootstrapper:78 - Launcher returned with code 1(0x1)
2023-05-11 13:32:22,708 INFO  [WrapperJarAppMain] DefaultAgentLauncherCreatorImpl:87 - Attempt No: 1 to cleanup launcher temp files
2023-05-11 13:32:22,795 INFO  [WrapperJarAppMain] AgentBootstrapper:108 - Waiting for 10000 ms before re-launch....
2023-05-11 13:32:32,912 INFO  [WrapperJarAppMain] DefaultAgentLauncherCreatorImpl:67 - Attempting to load com.thoughtworks.go.agent.launcher.AgentLauncherImpl as specified by manifest key Go-Agent-Launcher-Class
2023-05-11 13:32:32,923 INFO  [WrapperJarAppMain] ParentClassAccessFilteringClassloader:43 - Loading com.thoughtworks.cruise.agent.common.launcher.AgentLauncher using java.net.URLClassLoader
2023-05-11 13:32:32,947 INFO  [WrapperJarAppMain] AgentBootstrapper:74 - Attempting create and start launcher...
2023-05-11 13:32:32,952 INFO  [WrapperJarAppMain] ParentClassAccessFilteringClassloader:43 - Loading com.thoughtworks.cruise.agent.common.launcher.AgentLaunchDescriptor using java.net.URLClassLoader
```

You didn't include this relevant part of the error log earlier so I was guessing `Caused by: java.lang.ExceptionInInitializerError: Exception java.lang.UnsatisfiedLinkError: /tmp/jna434117078481342640.tmp: /tmp/jna434117078481342640.tmp: failed to map segment from shared object: Operation not permitted`

Are there permissions issues on your `/tmp`? Is your tmp mounted as `noexec` ? Could be similar issue as for https://www.elastic.co/guide/en/elasticsearch/reference/current/executable-jna-tmpdir.html

Hi @chadlwilson,
you are right about /tmp folder, there is noexec. This error line was missing in previous tries, maybe because of two versions of Java installed on the system. I've tried to use jna.tmpdir in the wrapper-properties:
```
wrapper.java.additional.100=-Djna.tmpdir=/var/tmp
```
I can see Java is running with that option now:
```
go        9555  4.2  0.7 9092032 119872 ?      Sl   11:26   0:05 /usr/lib/jvm/jdk-17-oracle-aarch64/bin/java -Dgocd.agent.log.dir=/var/log/go-agent -Djna.tmpdir=/var/tmp
```
But still facing same problem with /tmp:
```
Caused by: java.lang.ExceptionInInitializerError: Exception java.lang.UnsatisfiedLinkError: /tmp/jna13366009454418608986.tmp: /tmp/jna13366009454418608986.tmp: failed to map segment from shared object: Operation not permitted [in thread "main"]
```

I've removed the noexec from the /tmp and now go-agent starts. It's not ideal as I want this flag to the /tmp partition, but... 

Ok, once again - thank you and we can close the issue.

--- a/agent-bootstrapper/src/main/java/com/thoughtworks/go/agent/bootstrapper/LauncherTempFileHandler.java
+++ b/agent-bootstrapper/src/main/java/com/thoughtworks/go/agent/bootstrapper/LauncherTempFileHandler.java
@@ -27,20 +27,16 @@ import java.io.IOException;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
+import java.util.concurrent.TimeUnit;
 
 class LauncherTempFileHandler implements Runnable {
-
+    private static final Logger LOG = LoggerFactory.getLogger(LauncherTempFileHandler.class);
+    private static final String LAUNCHER_TMP_FILE_LIST = ".tmp.file.list";
     private static volatile Thread reaperThread;
-
-    static final String LAUNCHER_TMP_FILE_LIST = ".tmp.file.list";
-
-    private static final Logger LOG = LoggerFactory.getLogger(DefaultAgentLauncherCreatorImpl.class);
-
-    private static final int TEN_MINS = 1000 * 60 * 10;
 
     @Override
     public void run() {
-        while (true) {
+        while (!Thread.currentThread().isInterrupted()) {
             reapFiles();
             sleepForAMoment();
         }
@@ -83,8 +79,9 @@ class LauncherTempFileHandler implements Runnable {
 
     private static void sleepForAMoment() {
         try {
-            Thread.sleep(TEN_MINS);
+            Thread.sleep(TimeUnit.MINUTES.toMillis(10));
         } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
         }
     }
 }

--- a/agent/src/main/java/com/thoughtworks/go/agent/service/SystemInfo.java
+++ b/agent/src/main/java/com/thoughtworks/go/agent/service/SystemInfo.java
@@ -15,6 +15,7 @@
  */
 package com.thoughtworks.go.agent.service;
 
+import com.thoughtworks.go.util.FileUtil;
 import com.thoughtworks.go.util.SystemEnvironment;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -33,6 +34,7 @@ public class SystemInfo {
 
     static String determineOperatingSystemCompleteName() {
         try {
+            useAgentTmpDirIfNecessary();
             OperatingSystem os = newSystemInfo().getOperatingSystem();
             return String.format("%s %s%s",
                 os.getFamily(),
@@ -42,6 +44,12 @@ public class SystemInfo {
         } catch (Exception e) {
             LOG.warn("Unable to determine OS platform from native, falling back to default", e);
             return new SystemEnvironment().getOperatingSystemFamilyJvmName();
+        }
+    }
+
+    private static void useAgentTmpDirIfNecessary() {
+        if (System.getProperty("jna.tmpdir") == null) {
+            System.setProperty("jna.tmpdir", FileUtil.TMP_PARENT_DIR);
         }
     }
 

--- a/agent/src/test/java/com/thoughtworks/go/agent/service/SystemInfoTest.java
+++ b/agent/src/test/java/com/thoughtworks/go/agent/service/SystemInfoTest.java
@@ -16,17 +16,45 @@
 
 package com.thoughtworks.go.agent.service;
 
+import com.thoughtworks.go.util.FileUtil;
 import com.thoughtworks.go.util.SystemEnvironment;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.condition.EnabledOnOs;
 import org.junit.jupiter.api.condition.OS;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.io.TempDir;
 import org.mockito.Answers;
 import org.mockito.MockedStatic;
+import uk.org.webcompere.systemstubs.jupiter.SystemStub;
+import uk.org.webcompere.systemstubs.jupiter.SystemStubsExtension;
+import uk.org.webcompere.systemstubs.properties.SystemProperties;
+
+import java.nio.file.Path;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.Mockito.mockStatic;
 
+@ExtendWith(SystemStubsExtension.class)
 class SystemInfoTest {
+
+    @SystemStub
+    SystemProperties props;
+
+    @TempDir
+    Path tempDir;
+
+    @Test
+    public void shouldDefaultJnaTmpDirIfUnset() {
+        SystemInfo.determineOperatingSystemCompleteName();
+        assertThat(System.getProperty("jna.tmpdir")).isEqualTo(FileUtil.TMP_PARENT_DIR);
+    }
+
+    @Test
+    public void shouldPreserveJnaTmpDirIfSet() {
+        props.set("jna.tmpdir", tempDir.toString());
+        SystemInfo.determineOperatingSystemCompleteName();
+        assertThat(System.getProperty("jna.tmpdir")).isEqualTo(tempDir.toString());
+    }
 
     @Test
     @EnabledOnOs(OS.MAC)

